# Assignment 3: Selecting OPC Data For Review

## DSCI 504: Intro to SQL

Symphony Hopkins

September 6, 2022

**Step 1: Use a SELECT query to select all records from the customer table. Limit the displayed results to 10 records.**

In [12]:
SELECT c.cus_id
      ,cus_city
FROM dsci_504.customers AS c
LIMIT 10;

(10 row(s) affected)

Total execution time: 00:00:00.005

cus_id cus_city 2600 Hattiesburg 2601 Oswego 2602 Salt Lake City 2603 Coram 2604 Pataskala 2605 Southampton 2606 Los Angeles 2607 Cranston 2608 Ithaca 2609 Merrimack

**Q: Of the first ten customers returned, what is the Customer ID  and City of the 5th customer?**

```
A: The ID and city of the 5th customer is 2604 and Pataskala, respectively.

```

**Step 2: Use a SELECT query to select the Customer ID, Last Name, State, and Customer Appreciation Code of all customers who joined _from_ September 11, 2001 _through_ October 1, 2001**

In [19]:
SELECT cus_id
      ,cus_last_name
      ,cus_state
      ,cus_app_cd
FROM dsci_504.customers
WHERE cus_join_date
BETWEEN '2001-09-11' AND '2001-10-01';

(8 row(s) affected)

Total execution time: 00:00:00.004

cus_id cus_last_name cus_state cus_app_cd 356 Delacruz 2 6 464 Sherman 3 1 495 Delacruz 18 1 1587 Applewood 17 4 1907 Acosta 29 3 1936 Manchester 11 6 2141 Brooks 19 4 2194 Moore 28 4

**Q: How many customers joined in teh period from September 11, 2001 throught October 1, 2001?**

```
A: Eight customers joined in the period from September 11, 2001 and October 1, 2001.

```

**Part 3: Looking at the same query from Part 2, write a SELECT query that returns all information for the two customers with the last name Delacruz**

In [20]:

SELECT * 
FROM
(SELECT *
FROM dsci_504.customers
WHERE cus_join_date
BETWEEN '2001-09-11' AND '2001-10-01') AS q1
WHERE q1.cus_last_name = 'Delacruz';

(2 row(s) affected)

Total execution time: 00:00:00.003

cus_id cus_num cus_last_name cus_first_name cus_add_num cus_address cus_city cus_state cus_zip cus_phone cus_join_date cus_app_cd cus_app_num tot_ord_qty tot_ord_value 356 Del73426 Delacruz Margo 63740 Crystal Spring Corners Mcarthur Estates 2 89443 3875686732 2001-09-25 6 OPC80718306 70 34773 495 Del44118 Delacruz Brandon 53556 Apache Wolfs Mill 18 18151 7593410556 2001-09-23 1 OPC79747551 85 2396

**Q: What are three unique features about these two customers?**

```
They live in different states: Margo lives in Alabama and Brandon lives in North Carolina. They order a lot of things from OPC, so one can assume they are loyal customers. When you look at the total order quantity and total order value, we can see that Margo orders more expansive items than Brandon.

```

**Part 4: Write a query to return all order data for the two Delacruz customers above. Be sure to include minimal person information to assist in identifying which person placed which order. You may need to use a simple join to get both the customer and order data. A template for that is below. Use the following data points in your query:**

- **Customer ID**
- **Customer Last Name**
- **Customer First Name**
- **Total Order Quantity**
- **Total Order Value**
- **All Order Data**

```
SELECT [table1.column], [table2.column]..
FROM [table1], [table2]
WHERE [conditions]

```

In [70]:
SELECT *
FROM
(SELECT c.cus_id 
      ,c.cus_last_name
      ,c.cus_first_name
      ,c.tot_ord_qty
      ,c.tot_ord_value
FROM dsci_504.customers AS c
WHERE c.cus_id = '356'
OR c.cus_id = '495') AS q1
INNER JOIN dsci_504.orders AS o 
ON q1.cus_id = o.cus_id;

(2 row(s) affected)

Total execution time: 00:00:00.004

cus_id cus_last_name cus_first_name tot_ord_qty tot_ord_value ord_id ord_date ord_tax_loc order_tot ord_ship_add ord_ship_date ord_track_num prod_id warehouse_id cus_id 356 Delacruz Margo 70 34773 2614702 2002-08-21 2 5200 false 2002-08-25 PAC7774446-3 42 3 356 495 Delacruz Brandon 85 2396 2051970 2005-07-28 18 5631.35 false 2005-07-30 PAC2382180-2 48 1 495

**Q: What is one interesting fact revealed in the addition of the new data?**

```
A: The orders table does not include all of the orders each customer has made throughout the 20-year period.

```

**Q: Explain how the addition of additional data aided in your assessment.**

```
A: The order total (order_tot) and the number of sales transactions does not add up to the total order value (tot_ord_value) from the customers tables. 

```

**Part 5: Some of the most unique features mountain bikers seek out in their bikes are wheel sets, shocks, forks, and dropper posts. These items make or break a top- notch mountain bike. You know that part of the identification of whether or not OPC should break into the luxury MTB market is that they can build and market competitive bikes. Use a SELECT query to determine how many unique type of components there are in the components table. Specifically query for forks and dropper posts. Order your results by component category and component name in alphabetical order.**

In [29]:
SELECT *
FROM dsci_504.components as c
WHERE c.comp_cat IN ('fork','dropper')
GROUP BY c.comp_cat, c.comp_id
ORDER BY c.comp_cat, c.comp_name ASC;

(85 row(s) affected)

Total execution time: 00:00:00.004

comp_id comp_name comp_cost comp_supplier comp_cat sup_id 260 Bike Yoke Devine Stealth 168 37 dropper NULL 237 Bike Yoke Revive 315 37 dropper NULL 258 BrandX Ascend 125 42 dropper NULL 238 Cannondale 2 89 17 dropper NULL 245 Cannondale DownLow 89 17 dropper NULL 262 Cube Dropper 129 43 dropper NULL 236 DropWH18 78 36 dropper NULL 247 Fox Transfer 180 28 dropper NULL 259 Fox Transfer Factory 329 28 dropper NULL 234 FusionX Manic 189 34 dropper NULL 254 GT Dropkick 74 24 dropper NULL 243 Giant Contact Switch 189 1 dropper NULL 252 Ibis 90 22 dropper NULL 241 KS Lev 123 38 dropper NULL 248 KS Lev Integra 156 38 dropper NULL 256 KS Lev SI 134 38 dropper NULL 250 KS Rage-i 135 38 dropper NULL 249 Niner RDO 114 3 dropper NULL 244 OC2 78 61 dropper NULL 246 OC2 Carbon 199 61 dropper NULL 242 OnOff Pija 100 23 dropper NULL 253 OneUp V2 78 39 dropper NULL 235 Recon Dropper 113 35 dropper NULL 255 Recon Elite Dropper 113 35 dropper NULL 239 Ride Alpha 110 2 dropper NULL 240 RockShox Reverb Steakth 349 30 dropper NULL 257 SDG Tellis 250 40 dropper NULL 261 TransX 126.9 13 dropper NULL 251 X-Fusion Manix 199.96 73 dropper NULL 115 DVO Onyx D1 160 785 29 fork NULL 138 DVO Onyx D2 170 Air 790 29 fork NULL 136 Fox 32 Factory 100 489 28 fork NULL 142 Fox 32 Float Perf 100 Air 510 28 fork NULL 164 Fox 34 Float Factory Grip 2 120 969 28 fork NULL 150 Fox 34 Float Factory Grip 2 140 789 28 fork NULL 140 Fox 34 Float Factory Grip 2 140 Air 1039 28 fork NULL 128 Fox 34 Float Perf 120 429 28 fork NULL 153 Fox 34 Float Perf 140 410 28 fork NULL 162 Fox 34 Float Perf 140 Air 429 28 fork NULL 141 Fox 34 Perf 130 429 28 fork NULL 147 Fox 34 Perf 130 Air 429 28 fork NULL 113 Fox 34 Rhythm 120 429 28 fork NULL 135 Fox 36 Factory 150 Air 1069 28 fork NULL 134 Fox 36 Factory 160 Air 1069 28 fork NULL 155 Fox 36 Float Factory 160 Air 605 28 fork NULL 163 Fox 36 Float Factory 170 Air 999 28 fork NULL 124 Fox 36 Float Perf 140 Air 949 28 fork NULL 144 Fox 36 Float Perf 150 880 28 fork NULL 118 Fox 36 Float Perf 160 Air 849 28 fork NULL 143 Fox 36 Perf 160 Air 840 28 fork NULL 158 Fox 36 Perf 1650 Air 893 28 fork NULL 125 Fox 36 Perf 170 Air 798 28 fork NULL 165 Fox 36 Perf Elite 140 Air 789 28 fork NULL 146 Fox 36 Perf Elite 150 Air 899 28 fork NULL 166 Fox 36 Perf Elite 160 Air 789 28 fork NULL 148 Fox 36 Perf Elite 180 Air 1024 28 fork NULL 160 Fox 36 Rhythm 160 Air 556 28 fork NULL 116 Fox 36 Rhythm Fit 3 150 849 28 fork NULL 114 Fox 38 Factory Grip 2 180 Air 1249 28 fork NULL 123 Fox 38 Factory Grip2 170 Air 1249 28 fork NULL 132 Fox 38 Float Perf 170 Air 1020 28 fork NULL 119 Fox 38 Grip 2 170 Air 989 28 fork NULL 137 Lefty Ocho 100 Air 999 33 fork NULL 126 Marzocchi Bomber Z2 130 Air 610 32 fork NULL 159 Marzocchi Bomber Z2 Rail 120 Air 519 32 fork NULL 121 Ohlins RXF36 M.2 170 Air 1149.97 31 fork NULL 139 RockShox 35 Gold RL 140 Air 458 30 fork NULL 151 RockShox 35 Gold RL 150 Air 458 30 fork NULL 161 RockShox 35 Gold RL 160 Air 458 30 fork NULL 167 RockShox Lyric Sel 150 Air 765 30 fork NULL 130 RockShox Lyric Ult 150 Air 968 30 fork NULL 122 RockShox Lyric Ult RC2 160 968 30 fork NULL 120 RockShox Lyric Ult RC2 160 Air 968 30 fork NULL 149 RockShox Lyric Ult RC2 170 975 30 fork NULL 131 RockShox Pike Sel RC 140 713 30 fork NULL 152 RockShox Pike Sel+ 140 Air 713 30 fork NULL 157 RockShox Pike Ult 150 Air 917 30 fork NULL 145 RockShox Pike Ult RC 130 Air 917 30 fork NULL 117 RockShox Recon RL 120 274 30 fork NULL 133 RockShox Recon RL 120 Air 255 30 fork NULL 127 RockShox Recon RL 130 275 30 fork NULL 129 RockShox Recon RL 130 Air 270 30 fork NULL 154 RockShox YarRC 180 Air 509 30 fork NULL 168 RockShox YariRC 160 509 30 fork NULL 156 RockShox ZEB Ult 170 Air 915 30 fork NULL

**Q: How many unique combination of fork and dropper post were there in the data?**

```
A: There are 56 unique droppers and 29 unique forks.

```

**Part 6: Write a new query to identify all builds that the fork in row 60 is included on in the data. Include the accompanying build data in your query. Use a conditional query and search for the component name in the query.**

In [92]:
--For Part 6 - 8, I was unsure about the instructions because it seems like they more or less asking for the same thing.
SELECT *
FROM dsci_504.builds as b
WHERE b.comp_fork = '123';


(4 row(s) affected)

Total execution time: 00:00:00.003

build_id comp_hbar comp_brake comp_wheels comp_fork comp_shock comp_shifter comp_cassette comp_derail comp_dropper comp_crank comp_chain_ring build_name 71 43 67 91 123 184 61 219 273 260 273 273 precision 42 30 67 88 123 184 61 219 273 260 273 273 electric 45 26 67 72 123 192 273 274 229 259 273 273 megalodon 57 6 56 85 123 188 279 223 279 237 269 279 strikeforce

**Q: What happened in the query?**

```
A: The query returned all the build data that the fork is included in. From this query, we found that the fork is included in four builds.

```

**Part 7: Use the simple join concept from Part 4 to complete the query by joining the componet and build tables. Return all Build columns in your query, but not the component data.**

In [94]:
--I joined the component and build tables, and only returned the build data which is...
--why the query looks the same as Part 6.
SELECT q1.build_id
      ,q1.comp_hbar
      ,q1.comp_brake
      ,q1.comp_wheels
      ,q1.comp_fork
      ,q1.comp_shock
      ,q1.comp_shifter
      ,q1.comp_cassette
      ,q1.comp_derail 
      ,q1.comp_dropper
      ,q1.comp_crank
      ,q1.comp_chain_ring
      ,q1.build_name
FROM
(SELECT *
FROM dsci_504.builds) as q1
INNER JOIN dsci_504.components as c 
ON CAST(q1.comp_fork as INT) = c.comp_id
WHERE comp_fork = '123';


(4 row(s) affected)

Total execution time: 00:00:00.002

build_id comp_hbar comp_brake comp_wheels comp_fork comp_shock comp_shifter comp_cassette comp_derail comp_dropper comp_crank comp_chain_ring build_name 71 43 67 91 123 184 61 219 273 260 273 273 precision 42 30 67 88 123 184 61 219 273 260 273 273 electric 45 26 67 72 123 192 273 274 229 259 273 273 megalodon 57 6 56 85 123 188 279 223 279 237 269 279 strikeforce

**Q: How many builds is the 'Fox 38 Factory Grip2 170 Air' included in?**

```
A: Four.

```

**Part 8: Write a query to find the build names of the four builds from Part 7 above.**

In [95]:
SELECT q1.build_id
      ,q1.comp_fork
      ,q1.build_name
FROM
(SELECT *
FROM dsci_504.builds) as q1
INNER JOIN dsci_504.components as c 
ON CAST(q1.comp_fork as INT) = c.comp_id
WHERE comp_fork = '123';

(4 row(s) affected)

Total execution time: 00:00:00.007

build_id comp_fork build_name 71 123 precision 42 123 electric 45 123 megalodon 57 123 strikeforce

**Part 9: Identify the most expensive product of 2019. Since we haven't learned aggregate functions yet, please do not use the max() function. Use conditional statements to return a single row with the product with the highest order value**

In [39]:

SELECT o.ord_date
       ,o.prod_id 
       ,p.prod_name
       ,p.prod_price
       ,o.order_tot
FROM dsci_504.products AS p
INNER JOIN dsci_504.orders AS o
ON p.prod_id = o.prod_id
WHERE o.ord_date
BETWEEN '2019-01-01' AND '2019-12-31'
ORDER BY p.prod_price DESC
LIMIT 2;


(2 row(s) affected)

Total execution time: 00:00:00.002

ord_date prod_id prod_name prod_price order_tot 2019-12-03 24 Rail 7399 7813.34 2019-02-10 22 Wreckoning 7399 7694.96

**Q: What is the product name of the product and the price of the mountain bike with the highest prive ordered in 2019?**

```
A: There were two products with the highest price ordered in 2019. They were the Rail and Wreckoning, which were both $7399.

```

**Part 10: Execute three additional queries on tables you haven't used yet. These can be at your discretion.**

In [1]:
--What are the top 5 most returned components?
SELECT c.comp_id
      ,c.comp_name
      ,c.comp_cat
      ,SUM(r.tot_ret_item_cnt) AS tot_returns
FROM dsci_504.racs AS r
INNER JOIN dsci_504.components AS c 
ON r.comp_id = c.comp_id
GROUP BY c.comp_id
ORDER BY tot_returns DESC
LIMIT 5;



(5 row(s) affected)

Total execution time: 00:00:00.029

comp_id comp_name comp_cat tot_returns 56 Sram Code RSC brake 123 32 Ibis Carbon handlebar 104 41 Evil Boomstick 7075 Alloy handlebar 88 38 Specialized Trail 7050 Alloy handlebar 85 26 Phoenix Team LowRise handlebar 83

In [63]:
--How many orders does each warehouse recieve?
SELECT COUNT(cwo.warehouse_id) AS tot_warehouse_orders
FROM dsci_504.customerwarehouseorders AS cwo
GROUP BY cwo.warehouse_id;

(3 row(s) affected)

Total execution time: 00:00:00.002

tot_warehouse_orders 884 856 859

In [97]:
--Which state has the highest tax rate?
SELECT *
FROM dsci_504.taxes AS t 
ORDER BY t.tax_rate DESC
LIMIT 1;

(1 row(s) affected)

Total execution time: 00:00:00.002

tax_id tax_location tax_rate 17 MN 8.875

**Q: Write a few paragraphs about what you learned this week in the course. What interesting things did you discover in this assignment with simple queries?**

```
A: I knew a little bit about SQL before taking this course, but going through the modules this week has given me the oppurtunity to relearn some of things that I have learned previously. This week we went over the SQL execution order vs the written order, and that has helped me understand why errors arise when I try to execute certain queries. We also went over some of the SELECT query elements like WHERE, AND/OR, and ORDER BY, and reading those modules helped me remember aspects of each element. 

```
```
While working the assignments, some interesting things I discovered was that there are a lot of different components that make up a bike. I am not biker, so all of this is new to me; however, it was interesting to see how many different brands of a bike dropper and fork exists. The various components even inspired me to create a query about the top 5 most returned components (because surely not every component is well-produced or well-liked by the customers). Another interesting thing I found was that Minnesota had the highest sales tax rate. For some reason, I was expecting to see California or another high cost-of-living state on top.  

```